src/libsyntax/ast.rs

```diff
@@ -1389,14 +1389,15 @@ pub enum FunctionRetTy {
     /// raise an error or exit (i.e. never return to the caller)
     NoReturn(Span),
     /// Everything else
-    Return(P<Ty>),
+    Return(Option<P<Ty>>),
 }
 
 impl FunctionRetTy {
     pub fn span(&self) -> Span {
         match *self {
             NoReturn(span) => span,
-            Return(ref ty) => ty.span
+            Return(Some(ref ty)) => ty.span,
+            Return(None) => DUMMY_SP
```

Comment:
Could this possible lead to a regression in diagnostics? I'm not personally sure how often this `.span()` method is called, but it may be worth checking to make sure that it's basically never called if the function has no return type listed.